sourcegraph/sourcegraph
# gRPC: add support for accesslog interceptor in gitserver package

Closes #51056 

This PR adds gRPC support for the gitserver [access log](https://docs.sourcegraph.com/admin/audit_log) feature to bring it up to parity with the REST API. 

It does this by adding a few new components:
- https://github.com/sourcegraph/sourcegraph/blob/grpc-allowlist-accesslog/internal/requestclient/grpc.go: The AccessLog package relies on a `client` object being attached to the `context.Context` object for certain fields, including the IP address associated with the request. The HTTP middleware in the requestclient package accomplishes this by [using the `req.remoteAddr` field on the http request object](https://github.com/sourcegraph/sourcegraph/blob/4caec90feaea1143f5ccc1398707763fbbaa1e81/internal/requestclient/http.go#L77). I was able to mimic this in gRPC through the use of the of our [Propagator pattern](https://github.com/sourcegraph/sourcegraph/blob/grpc-allowlist-accesslog/internal/grpc/propagator/propagator.go).  If the client's IP / forwarded for address is already set in the request context (example: this gRPC call is the result of a graphQL request), then this propagator will correctly marshal / unmarshall those across the gRPC boundary. If the IP address isn't set in the client context (such as if the call is the result of some background job), the server will backfill the address via the [google.golang.org/grpc/peer](https://pkg.go.dev/google.golang.org/grpc/peer) package. This allows you to view the IP address information that gRPC server automatically attaches to the request context for every RPC call. 

	- This new client Propagation logic is part of the defaults for every gRPC server and client. 

- https://github.com/sourcegraph/sourcegraph/blob/grpc-allowlist-accesslog/cmd/gitserver/server/accesslog/accesslog.go: The `accesslog` package has been updated with new GRPC interceptors that inspect the request context after the handler finishes, and logs any fields if they were added via `Record`. 
	- These new interceptors are only added to the `gitserver` gRPC server, since the accesslog package as a whole has some hard-coded gitserver specific logic anyway. 
	- The existing HTTP accesslog logic only operates on certain methods. I mimicked this in GRPC with some shim interceptors that wrap existing ones that are only invoked if the method matches. See https://github.com/sourcegraph/sourcegraph/blob/4caec90feaea1143f5ccc1398707763fbbaa1e81/cmd/gitserver/shared/shared.go#L561-L585
		- I didn't think this would be a general utility that would be useful outside the gitserver package, so I didn't both exporting them.  


## Test plan

Two levels: added unit tests and manual testing. 

### manual testing

1. Enable access locally by using the information from https://docs.sourcegraph.com/admin/audit_log to enable it in your local site configuration: 
	```jsonc
	  "log": {
	    "auditLog": {
	      "internalTraffic": false,
	      "graphQL": false,
	      "gitserverAccess": true
	    }
	  }
	```

1. While you're at it, make sure that you have gRPC enabled locally:

	```
	 "experimentalFeatures": {
	    "enableGRPC": true
	  },
	```
1. See the following log lines appear in your terminal:

	```
	[    gitserver-0] INFO gitserver.exec.accesslog accesslog/accesslog.go:114 access logging enabled
	[    gitserver-0] INFO gitserver.archive.accesslog accesslog/accesslog.go:114 access logging enabled
	```

1. Perform a search that's likely to hit the `Exec` endpoint for gitserver, such as this symbol search: https://sourcegraph.test:3443/search?q=context%3Aglobal+index%3Ano+type%3Asymbol+test&patternType=standard&sm=1&groupBy=repo, and see similar log lines appear in your terminal 

	```
	[    gitserver-0] INFO gitserver.exec.accesslog audit/audit.go:43 access (sampling immunity token: c87c11c0-630b-4507-af3c-0596795e7a37) {"audit": {"auditId": "c87c11c0-630b-4507-af3c-0596795e7a37", "entity": "gitserver", "actor": {"actorUID": "3", "ip": "127.0.0.1:65068", "X-Forwarded-For": ""}}, "params": {"repo": "github.com/sourcegraph/sourcegraph", "cmd": "ls-tree", "args": ["--long", "--full-name", "-z", "9cee218369d2b292439bed60b2e4a94aae19d983", "--", "client/branded/src/components/panel/views/locations.test.ts"]}}
	[    gitserver-0] INFO gitserver.exec.accesslog audit/audit.go:43 access (sampling immunity token: 2a8dee8c-e513-4d37-8c81-a64ba39d2717) {"audit": {"auditId": "2a8dee8c-e513-4d37-8c81-a64ba39d2717", "entity": "gitserver", "actor": {"actorUID": "3", "ip": "127.0.0.1:65068", "X-Forwarded-For": ""}}, "params": {"repo": "github.com/sourcegraph/sourcegraph", "cmd": "ls-tree", "args": ["--long", "--full-name", "-z", "9cee218369d2b292439bed60b2e4a94aae19d983", "--", "client/branded/src/components/panel/views/locations.test.ts"]}}
	```

	_we only check the `Exec` endpoint since that's the only currently implemented gRPC endpoint that should accesslogging, see the following for reference_: 
	https://github.com/sourcegraph/sourcegraph/blob/1e1eb3fce1a2a8e9fce24a667b53105c2fe9a52e/cmd/gitserver/server/server.go#L558-L561
